### PR TITLE
[ENH] Venn Diagram: Add relations over columns, simplify over rows

### PR DESCRIPTION
**Pull request by AndrejaKovacic**
##### Issue

Implements  #3991. 

##### Description of changes
Venn diagram over rows is now simpler. Venn diagram over columns is now possible. I would appreciate comments about output, especially when using columns.

##### Includes
- [X] Code changes
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: Orange/widgets/visualize/owvenndiagram.py [hunk 1]
```diff
@@ -151,69 +170,81 @@ def __init__(self):
...
+        self.Error.domain_mismatch.clear()
+        self.Error.instances_mismatch.clear()
+        if not self.usecols:
+            domains = [input.table.domain for input in self.data.values()]
+            samedomain = all((d1 == d2) for d1, d2 in pairwise(domains))
+            if not samedomain:
+                self.vennwidget.clear()
+                self.Error.domain_mismatch()
+                self.itemsets = OrderedDict()
+                return False
```
Comment (janezd) [inline Orange/widgets/visualize/owvenndiagram.py]: These four lines repeat again below. You can extract this into a function (including `return False`) and replace these lines with `return name_of_that_function()`.

File: Orange/widgets/visualize/owvenndiagram.py [hunk 2]
```diff
@@ -105,27 +125,26 @@ def __init__(self):
...
-            model = itemmodels.VariableListModel(parent=self)
-            cb = gui.OrangeComboBox(
-                minimumContentsLength=12,
-                sizeAdjustPolicy=QComboBox.AdjustToMinimumContentsLengthWithIcon
+        box = gui.vBox(self.inputsBox, 'feature', addSpace=False)
+        box.setFlat(True)
+        model = itemmodels.VariableListModel(parent=self)
+        self.cb = cb = gui.OrangeComboBox(
+            minimumContentsLength=12,
+            sizeAdjustPolicy=QComboBox.AdjustToMinimumContentsLengthWithIcon
```
[the remark this answers is not in the document]
Comment (AndrejaKovacic) [reply Orange/widgets/visualize/owvenndiagram.py]: I agree, this is clearer. The referenced issue in this pull request has a sketch by @BlazZupan with "Use identifiers", but i think that's because of historical reasons.

**Comment by janezd** [comment]
TL;DR: Checkout my commit and see what I've done. :|

I'm sorry... There was a huge gap between the last radio button and the combo. I tried removing it and noticed there's an unnecessary box. And that the widget uses attribute indices althought the combo box uses a model. After fixing this, I realized that radio buttons are not necessary if we just use "Instance identity" as a placeholder for None. So, changing this, I noticed that the combo is disabled unless the diagram shows rows, hence, the combo actually belongs to one of the options in the other radio button group. This made the control area almost empty, in particular when I removed the empty and useless info box. So we arrived to what I've just committed.

**Comment by janezd** [comment]
Screenshot 2019-11-14 at 23 09 57

**Comment by janezd** [comment]
@AndrejaKovacic, in your last commit you removed context settings. You decided you didn't like them? :)

(I've modified the commit to put them back.)

File: Orange/widgets/visualize/owvenndiagram.py [hunk 3]
```diff
@@ -479,41 +344,21 @@ def _on_selectionChanged(self):
...
+        self.selection = [i for i, area in enumerate(areas) if area.isSelected()]
         self.invalidateOutput()
 
-    def _on_useidentifiersChanged(self):
-        self.inputsBox.setEnabled(self.useidentifiers == 1)
-        # Invalidate all itemsets
-        self._invalidate()
-        self._updateItemsets()
+    def _on_matching_changed(self):
+        self.output_duplicates_box.setEnabled(bool(self.rowwise))
```
Comment (lanzagar) [inline Orange/widgets/visualize/owvenndiagram.py]: Errrr, I don't get why the whole Output box would be disabled for Columns - including autocommit!
Probably only the `Output duplicates` option should be?

Edit: maybe this bug also hints at non-optimal naming - `output_duplicates_box` could be just `output_box` (duplicates setting is only one part of it).

File: Orange/widgets/visualize/owvenndiagram.py [hunk 4]
```diff
@@ -137,221 +100,163 @@ def __init__(self):
...
+        return False
+
+    def settings_compatible(self):
+        self.Error.domain_mismatch.clear()
+        self.Error.instances_mismatch.clear()
+        if self.rowwise:
+            domains = [input.table.domain for input in self.data.values()]
+            self.samedomain = all((d1 == d2) for d1, d2 in pairwise(domains))
+            if not self.samedomain:
+                return self.settings_incompatible()
```
Comment (lanzagar) [inline Orange/widgets/visualize/owvenndiagram.py]: The domains do not need to match. That is why there is an option to change `Rows, matched by:` from `Instance identity` (where it makes sense to match domains first) to a single variable with unique identifiers (matching is done only on that attribute, the rest of the domains can differ).

To test use the zoo dataset, connect to Venn. Also connect to PCA->Scatter plot, select some points and connect to Venn. Venn should be able to match on attribute `Name`.

File: Orange/widgets/visualize/owvenndiagram.py [hunk 5]
```diff
@@ -687,7 +596,7 @@ def __hash__(self):
     def __eq__(self, other):
         # XXX: comparing NaN with different payload
         return (isinstance(other, ComparableInstance)
-                and domain_eq(self.domain, other.domain)
+                and (self.domain == other.domain)
```
Comment (lanzagar) [inline Orange/widgets/visualize/owvenndiagram.py]: This now compares metas as well (before we didn't). Intentional?

**Comment by lanzagar** [comment]
The mentioned problems seem to be fixed. I found a new one, and the documentation update is still needed (can be done last).
- [ ] Instance identity used to ignore metas, now it uses them as well (can cause problems when a Selected column was added etc.). I don't think this was intentional?
- [ ] Update documentation

File: Orange/widgets/visualize/owvenndiagram.py (hunk 5 above)
Comment (AndrejaKovacic) [reply Orange/widgets/visualize/owvenndiagram.py]: It was requested in the original issue, that domains match completely, when doing comparisons over rows. This is a remnant of that, but as we later decided against the idea, i should have  reverted it. Good catch.
